File: dotnet/Base/OpenStack.Polyfills/System.Collections.Generic/GenericPool.cs
using System.Collections.Concurrent;
using System.Threading.Tasks;

namespace System.Collections.Generic
{
    public delegate void GenericPoolAction<T>(Action<T> action);
    public delegate TResult GenericPoolFunc<T, TResult>(Func<T, TResult> action);

    public interface IGenericPool<T>
    {
        T Get();
        void Release(T item);
        void Action(Action<T> action);
        TResult Func<TResult>(Func<T, TResult> action);
        Task ActionAsync(Func<T, Task> action);
        Task<TResult> FuncAsync<TResult>(Func<T, Task<TResult>> action);
    }

    public class GenericPool<T>(Func<T> factory, Action<T> reset = null, int retainInPool = 10) : IGenericPool<T>, IDisposable where T : IDisposable
    {
        readonly ConcurrentBag<T> items = [];
        public readonly Func<T> Factory = factory;
        public readonly int RetainInPool = retainInPool;

        public void Dispose()
        {
            foreach (var item in items) item.Dispose();
        }

        public virtual T Get() => items.TryTake(out var item) ? item : Factory();

        public virtual void Release(T item)
        {
            if (items.Count < RetainInPool) { reset?.Invoke(item); items.Add(item); }
            else item.Dispose();
        }

        public void Action(Action<T> action)
        {
            var item = Get();
            try { action(item); }
            finally { Release(item); }
        }

        public TResult Func<TResult>(Func<T, TResult> action)
        {
            var item = Get();
            try { return action(item); }
            finally { Release(item); }
        }

        public Task ActionAsync(Func<T, Task> action)
        {
            var item = Get();
            try { action(item); return Task.CompletedTask; }
            finally { Release(item); }
        }

        public Task<TResult> FuncAsync<TResult>(Func<T, Task<TResult>> action)
        {
            var item = Get();
            try { return action(item); }
            finally { Release(item); }
        }
    }

    public class SinglePool<T>(T single, Action<T> reset = null) : GenericPool<T>(null, reset) where T : IDisposable
    {
        public readonly T Single = single;
        public override T Get() => Single;
        public override void Release(T item) { Single.Dispose(); }
    }

    public class StaticPool<T>(T @static, Action<T> reset = null) : GenericPool<T>(null, reset) where T : IDisposable
    {
        public readonly T Static = @static;
        public override T Get() => Static;
        public override void Release(T item) { reset?.Invoke(item); }
    }
}

File: Python/openstk/core/genericpool.py
# https://stackoverflow.com/questions/74472798/how-to-define-python-generic-classes
from typing import Generic, TypeVar

T = TypeVar('T')

class IGenericPool(Generic[T]):
    def get(self) -> None: pass
    def release(self, item: T) -> None: pass
    def action(self, action: callable) -> None: pass
    def func(self, action: callable) -> object: pass

class GenericPool(Generic[T]):
    def __init__(self, factory: callable, reset: callable = None, retainInPool: int = 10):
        self.items: list[T] = []
        self.factory: callable = factory
        self.reset: callable = reset
        self.retainInPool: int = retainInPool
    def __enter__(self): return self
    def __exit__(self, *args):
        for s in self.items: s.__exit__(*args)
    
    def get(self) -> None:
        return self.items.pop() if self.items else self.factory()

    def release(self, item: T) -> None:
        if len(self.items) < self.retainInPool: self.reset and self.reset(item); self.items.append(item)
        else: item.__exit__(None, None, None)

    def action(self, action: callable) -> None:
        item = self.get()
        try: action(item)
        finally: self.release(item)

    def func(self, action: callable) -> object:
        item = self.get()
        try: return action(item)
        finally: self.release(item)

class SinglePool(GenericPool, Generic[T]):
    def __init__(self, single: T, reset: callable = None): super().__init__(None, reset); self.single: T = single
    def get(self) -> None: return self.single
    def release(self, item: T) -> None: self.reset and self.reset(item); self.single.__exit__(None, None, None)

class StaticPool(GenericPool, Generic[T]):
    def __init__(self, single: T, reset: callable = None): super().__init__(None, reset); self.static: T = static
    def get(self) -> None: return self.static
    def release(self, item: T) -> None: self.reset and self.reset(item)